objects/src/block/header.rs

```diff
@@ -275,17 +272,111 @@ impl Deserializable for BlockHeader {
     }
 }
 
-// UTILITIES
+// BLOCK NUMBER
 // ================================================================================================
 
-/// Returns the block number of the epoch block for the given `epoch`.
-pub const fn block_num_from_epoch(epoch: u16) -> u32 {
-    (epoch as u32) << BlockHeader::EPOCH_LENGTH_EXPONENT
+/// A convenience wrapper around a `u32` representing the number of a block.
+///
+/// Each block has a unique number and block numbers increase monotonically by `1`.
+#[derive(Debug, Default, Eq, PartialEq, Copy, Clone, PartialOrd, Ord, Hash)]
+pub struct BlockNumber(u32);
+
+impl BlockNumber {
+    /// The length of an epoch expressed as a power of two. `2^(EPOCH_LENGTH_EXPONENT)` is the
+    /// number of blocks in an epoch.
+    ///
+    /// The epoch of a block can be obtained by shifting the block number to the right by this
+    /// exponent.
+    pub const EPOCH_LENGTH_EXPONENT: u8 = 16;
+
+    /// Returns the previous block number
+    pub fn parent(&mut self) -> u32 {
+        self.0 - 1
+    }
+    /// Returns the next block number
+    pub fn child(&mut self) -> u32 {
+        self.0 + 1
+    }
+
+    /// Creates the [`BlockNumber`] corresponding to the epoch block for the provided `epoch`.
+    pub const fn from_epoch(epoch: u16) -> BlockNumber {
+        BlockNumber((epoch as u32) << BlockNumber::EPOCH_LENGTH_EXPONENT)
+    }
+
+    /// Creates a `BlockNumber` from a `usize`.
+    pub fn from_usize(value: usize) -> Self {
+        BlockNumber(value as u32)
+    }
+
+    /// Returns the epoch to which this block number belongs.
+    pub const fn block_epoch(&self) -> u16 {
+        (self.0 >> BlockNumber::EPOCH_LENGTH_EXPONENT) as u16
+    }
+
+    /// Returns the block number as a `u32`.
+    pub fn as_u32(&self) -> u32 {
+        self.0
+    }
+
+    /// Returns the block number as a `u64`.
+    pub fn as_u64(&self) -> u64 {
+        self.0 as u64
+    }
+
+    /// Returns the block number as a `usize`.
+    pub fn as_usize(&self) -> usize {
+        self.0 as usize
+    }
+}
+
+impl Add for BlockNumber {
+    type Output = Self;
+
+    fn add(self, other: Self) -> Self::Output {
+        BlockNumber(self.0 + other.0)
+    }
 }
 
-/// Returns the epoch of the given block number.
-pub const fn block_epoch_from_number(block_number: u32) -> u16 {
-    (block_number >> BlockHeader::EPOCH_LENGTH_EXPONENT) as u16
+impl Sub for BlockNumber {
+    type Output = Self;
+
+    fn sub(self, other: Self) -> Self::Output {
+        BlockNumber(self.0 - other.0)
+    }
+}
```

Comment:
I'd remove this as we would probably want checked subtraction. So instead make this:


```rust
/// Checked integer subtraction. Computes `self - rhs`, returning `None` if overflow occurred.
pub fn checked_sub(&self, rhs: u32) -> Option<Self> {
    self.0.checked_sub(rhs).map(Self)
}
```